### PR TITLE
Replace Resampler macros with implementations generic over a Sample

### PR DESCRIPTION
**Pull request by udoprog**
Note: this is built on top of #13, so it should be merged before this. This also conflicts w/ #12 - but assuming this is accepted I'd happily rework either of those PRs.

This removes the macros responsible for building sampler implementations w/ generic sampler implementations over a `Sample` trait (which is quite wordy). And since we're now generic over the `Sample` trait, the vectorised interpolation support also needed to be reworked.

Each kind of vectorized operation has their own `Sample` trait (ex `SseSample`). Which defines all the necessary operations in order to operate over a specific kind of sample. This has at least one benefit in that it allows us to remove the `PhantomData` associated with all the interpolators and use a single generic `sincs` field.

All of these traits are implemented by the `Sample` trait, and if a platform is not supported the trait is replaced by a macro with a dummy implementation that does nothing. In that sense, it's like an optional supertrait.

The feature detection has also been moved into a single place. The constructor of the interpolator kind being built. And feature errors are now runtime errors of the kind `MissingCpuFeatures` which describes exactly which features are missing. This can be either be formatted as an error, or simply unwrapped to get the same level of diagnostics as the panic.

Benefits include:
* Improved IDE and formatter support. Completions, jump to definitions, and so forth are all a bit wonky when the entire impl is inside of a macro.
* A user of the library can be generic over the `Sample` type.

The `Sample` trait is [a bit wordy to say the least with a lot of super traits](https://github.com/HEnquist/rubato/pull/15/files#diff-a7c5a2c1343c75492c1665b808866f0f41f45b9b7d4ac52f0694901b3e8a26eaR8)! This can in my estimate be reduced quite heavily be removing a large amount of repeated code in the future. But I decided to take this route right now to reduce the amount of refactoring done in a single patch. I'm personally not bothered by this. But it does mean that for each kind of operation added work needs to be done to the `Sample` trait.

Finally, I've modified the build configuration to build on all platforms supported by Github Actions for free. I've also removed the `check` build stage, because it's superfluous when doing the other build steps.

### CODE REVIEW

**Comment by udoprog** [comment]
I also noticed one minor change I've been working on slipped through that I didn't intend to. Instead of having updating an async sampler be a runtime error it is now part of the trait, so there's two kinds of resamplers:
* `Resampler`
* `AsyncResampler` - which has the `set_*` methods.

I intended to put it in a separate PR, but hey. I'll remove it if it's not wanted here.

**Comment by udoprog** [comment]
Oh, and I don't have an aarch64 environment I can test this on. So not sure if the neon impl still compiles :)

File: .github/workflows/ci_test.yml [hunk 1]
```diff
@@ -3,34 +3,12 @@ on: [push, pull_request]
...
-      - name: Run cargo check nodefault
-        uses: actions-rs/cargo@v1
-        with:
-          command: check
-
-
   test:
     name: Test Suite
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
```
Comment (HEnquist) [inline .github/workflows/ci_test.yml:8]: I know it's free to run them all, but this just feels like wasting resources. There is nothing os-specific in this library. What would make sense would be to run on different cpu architectures (something I planned to add but didn't get around to yet).

File: .github/workflows/ci_test.yml [hunk 2]
```diff
@@ -3,34 +3,12 @@ on: [push, pull_request]
...
-        uses: actions/checkout@v2
-
-      - name: Install stable toolchain
-        uses: actions-rs/toolchain@v1
-        with:
-          profile: minimal
-          toolchain: stable
-          override: true
-
-      - name: Run cargo check
```
Comment (HEnquist) [inline .github/workflows/ci_test.yml:20]: I think the check command is still useful and should stay. It runs much quicker than for example test, and you can quickly try all combinations of features.

File: src/interpolator_avx.rs [hunk 3]
```diff
@@ -227,8 +216,12 @@ mod tests {
         let oversampling_factor = 256;
         let window = WindowFunction::BlackmanHarris2;
         let sincs = make_sincs::<f64>(sinc_len, oversampling_factor, f_cutoff, window);
-        let interpolator =
-            AvxInterpolator::<f64>::new(sinc_len, oversampling_factor, f_cutoff, window);
+
+        let interpolator = match AvxInterpolator::<f64>::new(sinc_len, oversampling_factor, f_cutoff, window) {
+            Ok(interpolator) => interpolator,
+            Err(..) => return,
```
Comment (HEnquist) [inline src/interpolator_avx.rs:222]: This means it can get away with an error. If new returns an error, the test will pass.

**Comment by HEnquist** [comment]
> I also noticed one minor change I've been working on slipped through that I didn't intend to. Instead of having updating an async sampler be a runtime error it is now part of the trait, so there's two kinds of resamplers:
> 
>     * `Resampler`
> 
>     * `AsyncResampler` - which has the `set_*` methods (link: https://github.com/HEnquist/rubato/pull/15/files#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R205).
> 
> 
> I intended to put it in a separate PR, but hey. I'll remove it if it's not wanted here.

In the application I mainly wrote this for, it's much easier if there is a single resampler type. The `set_` methods are useless on the synchronous one yes, but having two different resampler traits means I will need to add a lot of extra stuff to handle that.

File: src/interpolator_neon.rs [hunk 4]
```diff
@@ -225,6 +156,61 @@ impl NeonInterpolator<f64> {
...
+        if !is_aarch64_feature_detected!("neon") {
+            return Err(MissingCpuFeatures(FEATURES));
+        }
+
+        assert!(sinc_len % 8 == 0, "Sinc length must be a multiple of 8.");
+        let sincs = make_sincs(sinc_len, oversampling_factor, f_cutoff, window);
+        let sincs = unsafe { <T as NeonSample>::pack_sincs(sincs) };
+
+        Ok(Self {
+            sinc,
```
Comment (HEnquist) [inline src/interpolator_neon.rs:206]: typo, should be "sincs"

File: src/interpolator_neon.rs [hunk 5]
```diff
@@ -225,6 +156,61 @@ impl NeonInterpolator<f64> {
...
+            length: sinc_len,
+            nbr_sincs: oversampling_factor,
+            phantom: PhantomData,
```
Comment (HEnquist) [inline src/interpolator_neon.rs:209]: remove "phantom". This and the sinc typo prevent it from compiling with neon support.

File: src/asynchro.rs [hunk 6]
```diff
@@ -139,556 +139,563 @@ pub struct SincFixedOut<T> {
...
+    }
+
+    #[cfg(all(target_arch = "aarch64", feature = "neon"))]
+    if let Ok(interpolator) =
+        NeonInterpolator::<T>::new(sinc_len, oversampling_factor, f_cutoff, window)
+    {
+        return Box::new(interpolator);
+    }
+
+    Box::new(ScalarInterpolator::<T>::new(
```
Comment (HEnquist) [inline src/asynchro.rs:180]: The benches are failing now because the simd interpolator constructors all return Result<Self, ..>, while the scalar one directly returns Self. The easiest fix is to let also that one return a Result, and then add an unwrap() in the benches.

File: src/interpolator_avx.rs (hunk 3 above)
Comment (udoprog) [reply src/interpolator_avx.rs:222]: Indeed. all though the only error it returns is if the avx interpolator isn't supported, which it isn't on the macos build agents. This was a quick and dirty way to disable (or just make the test pass) in case it's not available. I can clean it up a bit.

File: src/interpolator_neon.rs (hunk 4 above)
Comment (udoprog) [reply src/interpolator_neon.rs:206]: Yep!

Are you able to build on an aarch64 platforms w/ neon? I'm trying to set one up in a virtual machine, but I'm curious how you do it?

Comment (HEnquist) [reply src/interpolator_neon.rs:206]: It's quite easy with cross, then it runs in qemu which supports neon: https://github.com/rust-embedded/cross

But even easier is to run it on a raspberry pi with a 64bit os, that's what I usually do to test locally.

File: src/interpolator_avx.rs (hunk 3 above)
Comment (HEnquist) [reply src/interpolator_avx.rs:222]: you can add an assert in the Err arm, to check that avx really isn't supported. Check that with `is_x86_feature_detected!("avx")`

Comment (udoprog) [reply src/interpolator_avx.rs:222]: So it would be a duplicate of the check inside of the constructor. What I was considering replacing it with would be something like:
```rust
let interpolator = match AvxInterpolator::<f64>::new(sinc_len, oversampling_factor, f_cutoff, window) {
    Ok(interpolator) => interpolator,
    // expected in case test is run on a platform without all the necessary CPU features.
    Err(MissingCpuFeatures(&[CpuFeature::Avx, CpuFeature::Fma])) => return,
    Err(e) => panic!("{}", e),
};
```
That way, it'll panic if there's a mismatch in the expected feature, and the check is only done in one place.

But I'll add a feature check in the test if that's what you want! libtest doesn't support disabling tests at runtime to my knowledge, so the best we can do is the same as this currently does and return.

**Comment by HEnquist** [comment]
Overall I'm happy with these changes and I will merge once we sort out the few things I found.
Since there will be breaking changes, a new major version is needed. I think I'll skip to 1.0.0, that 0.x.y looks ugly.
There was a new version of num recently, and that will give new versions of several dependencies, including my own realfft. I think it makes sens to wait for all that to land before making a new rubato release. Hopefully doesn't take very long.

File: .github/workflows/ci_test.yml (hunk 1 above)
Comment (udoprog) [reply .github/workflows/ci_test.yml:8]: macos at least is useful since it's a common platform and it doesn't have the fma CPU feature. Ultimately it's a bit of an issue that CI isn't able to at least perform a positive test on hardware configuration (e.g. neon is not covered) which can easily let errors slip by.

You want me to nuke this?

**Comment by udoprog** [comment]
> In the application I mainly wrote this for, it's much easier if there is a single resampler type. The `set_` methods are useless on the synchronous one yes, but having two different resampler traits means I will need to add a lot of extra stuff to handle that.

I'll resurrect the error type if that's what you want. I didn't intend it to be in here anyways!

File: src/interpolator_avx.rs [hunk 7]
```diff
@@ -10,85 +11,44 @@ use core::arch::x86_64::{
...
-        self.length
-    }
-
-    fn nbr_sincs(&self) -> usize {
-        self.nbr_sincs
-    }
+use crate::Sample;
+
+/// Collection of cpu features required for this interpolator.
+static FEATURES: &[CpuFeature] = &[CpuFeature::Avx, CpuFeature::Fma];
```
Comment (HEnquist) [inline src/interpolator_avx.rs:18]: This could be simplified. There has ever been a cpu with fma, that doesn't also have avx. And since we require fma, it's enough to report if that is missing. If avx isn't supported, fma isn't either. Then the list of features isn't needed. 

Anyway right now it's returning that both are missing even if avx is available, on for example a sandy bridge that has avx but not fma.

**Comment by HEnquist** [comment]
> > In the application I mainly wrote this for, it's much easier if there is a single resampler type. The `set_` methods are useless on the synchronous one yes, but having two different resampler traits means I will need to add a lot of extra stuff to handle that.
> 
> I'll resurrect the error type if that's what you want. I didn't intend it to be in here anyways!

Resampler trait you mean?

**Comment by udoprog** [comment]
> Resampler trait you mean?

Yeah. But I meant the error that was returned in case it's not supported. It's done in the last patch.

**Comment by udoprog** [comment]
So lot's of clippy lints that came alive when I switched from `num_traits::Float::from` to `T::coerce(..)`. I'll fix them.

File: src/interpolator_avx.rs (hunk 7 above)
Comment (udoprog) [reply src/interpolator_avx.rs:18]: Yeah, it was just intended to do more-or-less the same thing in case you unwrap:ed the error as the panic previously did. We can definitely make it smarter and inform you of the exact sequence of features missing. All though then the error would have to be dynamic.

I'm not sure about feature flags tbh. Recent rustc experience has shaken my belief in how stable the assumptions are from cpu to cpu. But I'll defer to what you want to do here.

What would you want to change?

File: src/interpolator_avx.rs (hunk 3 above)
Comment (HEnquist) [reply src/interpolator_avx.rs]: I was thinking something simpler, like this:
```
let interpolator = match AvxInterpolator::<f64>::new(sinc_len, oversampling_factor, f_cutoff, window) {
    Ok(interpolator) => interpolator,
    Err(_) => {
        assert!(!(is_x86_feature_detected!("avx") && is_x86_feature_detected!("fma"));
        return;
    },
};
```
Then the test doesn't depend on MissingCpuFeatures, just a simple check that it was asslowed to return an error. Right now it's a duplicate of the check in new(), but that one could change and this simpler test would still be correct.

Comment (udoprog) [reply src/interpolator_avx.rs]: Done

File: src/interpolator_neon.rs (hunk 4 above)
Comment (udoprog) [reply src/interpolator_neon.rs]: Well, if you can run the neon tests I'd appreciate it for now (you can also push fixes to my branch directly making it easier). I don't have access to qemu right now!

File: .github/workflows/ci_test.yml (hunk 2 above)
Comment (udoprog) [reply .github/workflows/ci_test.yml:20]: Sure, I'll add it back if you want it. All though note that I believe clippy and check does roughly the same things right now (and both take ~1 minute to run).

I think `args: --no-default-features` is missing below which is what confused me, so I'll add that too.

File: .github/workflows/ci_test.yml (hunk 1 above)
Comment (HEnquist) [reply .github/workflows/ci_test.yml:39]: With the free github actions you don't know what cpu the job will run on. They have a very wide range of machines of different age, and it's really a lottery what you get. I wouldn't think all macos hosts lack fma (and I also wouldn't be sure all linux hosts have it). 

About arm and neon, that is possible to test with cross: https://github.com/ejmahler/RustFFT/blob/master/.github/workflows/run_test.yml#L63-L119
Those two tests are quite useful. 32-bit x86 checks that everything works as expected on a 32-bit machine, and since it runs without emulation it's still fast. And aarch64 can test with neon. That runs with qemu, so it's pretty slow. The tests here run pretty quickly so it probably won't be an issue. But if they take too long the tests can run be run with --release.

At some point there should also be tests running on several rustc versions. Let's drop windows and macos, to not make the list of runs overly long.

**Comment by HEnquist** [comment]
> > Resampler trait you mean?
> 
> Yeah. But I meant the error that was returned in case it's not supported. It's done in the last patch.

Ah alright, sure!

**Comment by HEnquist** [comment]
> So lot's of clippy lints that came alive when I switched from `num_traits::Float::from` to `T::coerce(..)`. I'll fix them if that's all right w/ you?

Sure! Nice to keep clippy happy.

File: .github/workflows/ci_test.yml (hunk 1 above)
Comment (udoprog) [reply .github/workflows/ci_test.yml]: Done

File: src/interpolator_neon.rs (hunk 4 above)
Comment (udoprog) [reply src/interpolator_neon.rs]: I managed to get it tested w/ cross and docker on my system. Seems to work now!

File: src/interpolator_avx.rs (hunk 7 above)
Comment (HEnquist) [reply src/interpolator_avx.rs:18]: The idea was that if we only care about fma, then the cpu features list will always contain only one element. And then it doesn't need to be a list. That in turn means that CpuFeature and MissingCpuFeatures could be merged to just MissingCpuFeature.
Of course when checking for fma, it should still also check for avx to make sure.

Comment (udoprog) [reply src/interpolator_avx.rs:18]: All righty, done!

**Comment by udoprog** [comment]
@HEnquist that's the last open comment I see which is open. Feel free to merge whenever you're comfortable and preferably squash. Or not ;)

This also closes #13 since this one includes it.

**Comment by HEnquist** [comment]
Looks good now! I will test it a bit more to make sure the resample results are all good. The tests only catch quite big problems, small issues can still sneak through. I don't expect any problems, and I'll merge as soon as I'm done.

**Comment by HEnquist** [comment]
I have run all my tests now and all the results look good. Merging!
